### PR TITLE
Exposed a new API to get number of CPU Cores on a Node

### PR DESCRIPTION
**Pull request by ninzavivek**
Exposed a new API to get number of CPU Cores on a Node. This will be used by DNC eventually to enforce per node NIC limit which is dependent on number of cores.

DNC before allocating a NIC, will get query a node for number of cores and see if the limit ha already been achieved,

### CODE REVIEW

File: cns/api.go [hunk 1]
```diff
@@ -18,6 +18,7 @@ const (
 	GetIPAddressUtilizationPath = "/network/ip/utilization"
 	GetUnhealthyIPAddressesPath = "/network/ipaddresses/unhealthy"
 	GetHealthReportPath         = "/network/health"
+	NumberOfCPUCoresPath        = "/network/hostcpucores"
```
Comment (ashvindeodhar) [inline cns/api.go]: Is there a need for hostcpucores to be under network?

File: cns/api.go [hunk 2]
```diff
@@ -139,6 +140,12 @@ type Response struct {
 	Message    string
 }
 
+// getNumberOfCPUCores describes reponse that returns the host local IP Address.
```
Comment (ashvindeodhar) [inline cns/api.go]: Fix the comment. (NumOfCPUCoresResponse)

File: cns/restserver/restserver.go [hunk 3]
```diff
@@ -177,6 +179,7 @@ func (service *HTTPRestService) Start(config *common.ServiceConfig) error {
 	listener.AddHandler(cns.V2Prefix+cns.DetachContainerFromNetwork, service.detachNetworkContainerFromNetwork)
 	listener.AddHandler(cns.V2Prefix+cns.CreateHnsNetworkPath, service.createHnsNetwork)
 	listener.AddHandler(cns.V2Prefix+cns.DeleteHnsNetworkPath, service.deleteHnsNetwork)
+	listener.AddHandler(cns.V2Prefix+cns.NumberOfCPUCoresPath, service.getNumberOfCPUCores)
```
Comment (ashvindeodhar) [inline cns/restserver/restserver.go]: I have never seen / understood the need of the V2Prefix. If we are not using it, let's get rid of this section? We can do this as a separate PR or this PR.

File: cns/restserver/restserver_test.go [hunk 4]
```diff
@@ -689,3 +689,23 @@ func TestGetInterfaceForNetworkContainer(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
```
Comment (ashvindeodhar) [inline cns/restserver/restserver_test.go]: Add a comment to describe the function

File: cns/restserver/restserver.go [hunk 5]
```diff
@@ -1573,3 +1576,32 @@ func (service *HTTPRestService) getNetPluginDetails() *networkcontainers.NetPlug
...
 	return networkcontainers.NewNetPluginConfiguration(pluginBinPath, configPath)
 }
+
+// Retrieves the number of logic processors on a node. It will be primarily
+// used to enforce per VM delegated NIC limit by DNC.
+func (service *HTTPRestService) getNumberOfCPUCores(w http.ResponseWriter, r *http.Request) {
+	log.Printf("[Azure CNS] getNumberOfCPUCores")
+	log.Request(service.Name, "getNumberOfCPUCores", nil)
+
+	var num = 0
```
Comment (ashvindeodhar) [inline cns/restserver/restserver.go]: Can you use the style - 
var (
num int,
...
)

Comment (ashvindeodhar) [reply cns/restserver/restserver.go]: GO inits the vars to 0

File: cns/restserver/restserver.go [hunk 6]
```diff
@@ -1573,3 +1576,32 @@ func (service *HTTPRestService) getNetPluginDetails() *networkcontainers.NetPlug
...
+	var returnCode = 0
+	var errmsg string
```
Comment (ashvindeodhar) [inline cns/restserver/restserver.go]: consistent naming style? errMsg?

File: cns/restserver/restserver_test.go [hunk 7]
```diff
@@ -689,3 +689,23 @@ func TestGetInterfaceForNetworkContainer(t *testing.T) {
...
+	fmt.Println("Test: getNumberOfCPUCores")
+
+	req, err := http.NewRequest(http.MethodGet, cns.NumberOfCPUCoresPath, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	w := httptest.NewRecorder()
+	mux.ServeHTTP(w, req)
+	var numOfCoresResponse cns.NumOfCPUCoresResponse
```
Comment (ashvindeodhar) [inline cns/restserver/restserver_test.go]: Declare vars upfront?
var (
)

File: cns/api.go [hunk 8]
```diff
@@ -140,7 +140,7 @@ type Response struct {
 	Message    string
 }
 
-// getNumberOfCPUCores describes reponse that returns the host local IP Address.
+// getNumberOfCPUCores describes reponse that returns num of cpu cores present on host.
```
Comment (ashvindeodhar) [inline cns/api.go]: Shouldn't the comment say - NumOfCPUCoresResponse describes ... ?

File: cns/restserver/restserver.go [hunk 9]
```diff
@@ -1573,3 +1576,32 @@ func (service *HTTPRestService) getNetPluginDetails() *networkcontainers.NetPlug
...
 	return networkcontainers.NewNetPluginConfiguration(pluginBinPath, configPath)
 }
+
+// Retrieves the number of logic processors on a node. It will be primarily
+// used to enforce per VM delegated NIC limit by DNC.
+func (service *HTTPRestService) getNumberOfCPUCores(w http.ResponseWriter, r *http.Request) {
+	log.Printf("[Azure-CNS] getNumberOfCPUCores")
+	log.Request(service.Name, "getNumberOfCPUCores", nil)
+
+	var num int
```
Comment (ashvindeodhar) [inline cns/restserver/restserver.go]: Can you combine the vars together ?
var (
num int
returnCode int
errMsg string
)

File: cns/restserver/restserver_test.go (hunk 4 above)
Comment (ashvindeodhar) [reply cns/restserver/restserver_test.go]: Can you respond to the review comment rather than marking it resolved directly if you are not going to act on it ?

File: cns/restserver/restserver_test.go [hunk 10]
```diff
@@ -689,3 +689,26 @@ func TestGetInterfaceForNetworkContainer(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func TestGetNumOfCPUCores(t *testing.T) {
+	fmt.Println("Test: getNumberOfCPUCores")
+
+	var err error
```
Comment (ashvindeodhar) [inline cns/restserver/restserver_test.go]: Same comment for var ( as above. 
Move var w *httptest.ResponseRecorder up front

File: cns/restserver/restserver_test.go (hunk 7 above)
Comment (ashvindeodhar) [reply cns/restserver/restserver_test.go]: Can you respond to the review comment rather than marking it resolved directly if you are not going to act on it ?

File: cns/api.go [hunk 11]
```diff
@@ -18,6 +18,7 @@ const (
 	GetIPAddressUtilizationPath = "/network/ip/utilization"
 	GetUnhealthyIPAddressesPath = "/network/ipaddresses/unhealthy"
 	GetHealthReportPath         = "/network/health"
+	NumberOfCPUCoresPath        = "/hostcpucores"
```
Comment (tamilmani1989) [inline cns/api.go]: can it be compute/hostcpucores?

File: cns/api.go (hunk 8 above)
Comment (ninzavivek) [reply cns/api.go]: test

File: cns/restserver/restserver_test.go (hunk 4 above)
Comment (ninzavivek) [reply cns/restserver/restserver_test.go]: Apparently, none of my responses got saved. I think Test name is self explanatory , hence no comment needed.

File: cns/restserver/restserver_test.go (hunk 10 above)
Comment (ninzavivek) [reply cns/restserver/restserver_test.go]: Nopes, recorder will not be needed if the new request itself fails.

File: cns/restserver/restserver.go (hunk 9 above)
Comment (ninzavivek) [reply cns/restserver/restserver.go]: :-). Is this being done anywhere in similar context?

File: cns/restserver/restserver.go (hunk 3 above)
Comment (ninzavivek) [reply cns/restserver/restserver.go]: I will like to keep it, as I don't know who all are using it.

File: cns/api.go (hunk 11 above)
Comment (ninzavivek) [reply cns/api.go]: let's not debate this to much. cns - container networking service.

Comment (tamilmani1989) [reply cns/api.go]: oki'm not forcing this. just a suggestion. cpucores ususally tie with compute.

File: cns/api.go (hunk 8 above)
Comment (tamilmani1989) [reply cns/api.go]: why did you resolve this?

File: cns/restserver/restserver.go (hunk 9 above)
Comment (tamilmani1989) [reply cns/restserver/restserver.go]: I agree its not done elsewhere but whatever ashvin suggested looks better. lets do this from now on
